File: Java-version/LeetCode/src/com/jooc/Common/Solution_357.java
package com.jooc.Common;

public class Solution_357 {
    public int countNumbersWithUniqueDigits_0(int n) {
        int[] count = new int[n+1 < 2 ? 2 : n+1];
        count[0] = 1;
        count[1] = 10;

        for(int i = 2; i <= n && i <= 10; i++){
            count[i] = 9;
            int num = 9;
            for(int j = i - 1; j > 0; j--){
                count[i] *= num--;
            }
            count[i] = count[i] + count[i-1];
        }

        return count[n];
    }

    public int countNumbersWithUniqueDigits(int n){
        if(n == 0) return 1;
        int pre = 1, cur = 10;

        for(int i = 1; i <= n && i <= 10; i++){
            cur = 9;
            int num = 9;
            for(int j = i - 1; j > 0; j--){
                cur *= num--;
            }
            cur = cur + pre;
            pre = cur;
        }

        return cur;
    }
}

File: python-version/leetcode/Solution_357.py
class Solution:
    def main(self):
        print(self.countNumbersWithUniqueDigits(3))
    def countNumbersWithUniqueDigits_0(self, n: int) -> int:
        if n == 0: return 1
        if n == 1: return 10
        ans, cur, factor = 10, 9, 9
        for _ in range(1, n):
            cur *= factor
            ans += cur
            factor -= 1

        return ans

    def countNumbersWithUniqueDigits(self, n: int) -> int:
        if n == 0: return 1
        if n == 1: return 10
        ans, cur = 10, 9
        for i in range(n-1):
            cur *= 9 - i
            ans += cur
        return ans

